src/Control/Monad/Eff/AVar.js

```diff
@@ -0,0 +1,318 @@
+    var cell = putLast(avar.reads, cb);
+    drainVar(left, right, avar);
+    return function () {
+      deleteCell(cell);
+    };
+  };
+};
+
+exports._tryPutVar = function (left, right, avar, value) {
+  return function () {
+    if (avar.value === EMPTY && avar.error === null) {
+      avar.value = value;
+      drainVar(left, right, avar);
+      return true;
+    } else {
+      return false;
+    }
+  };
+};
+
+exports._tryTakeVar = function (left, right, nothing, just, avar) {
+  return function () {
+    var value = avar.value;
+    if (value === EMPTY) {
+      return nothing;
+    } else {
+      avar.value = EMPTY;
+      drainVar(left, right, avar);
+      return just(value);
+    }
+  };
+};
+
+exports._tryReadVar = function (nothing, just, avar) {
+  return function () {
+    if (avar.value === EMPTY) {
+      return nothing;
+    } else {
+      return just(avar.value);
+    }
+  };
+};
+
+exports.isEmptyVar = function (avar) {
+  return function () {
+    return avar.value === EMPTY;
+  };
+};
+
+function drainVar (left, right, avar) {
+  if (avar.draining) {
+    return;
+  }
+
+  var ps = avar.puts;
+  var ts = avar.takes;
+  var rs = avar.reads;
+  var p, r, t, value, rsize;
+
+  avar.draining = true;
+
+  /* jshint -W084 */
+  while (1) {
+    p = null;
+    r = null;
+    t = null;
+    value = avar.value;
+    rsize = rs.size;
+
+    if (avar.error !== null) {
+      value = left(avar.error);
+      // Error callback ordering is somewhat undefined, but we try to at least
+      // be fair by interleaving puts and takes.
+      while (1) {
+        if (ps.size === 0 && ts.size === 0 && rs.size === 0) {
+          break;
+        }
+        if (p = takeHead(ps)) {
+          runEff(p.cb(value));
+        }
+        while (r = takeHead(rs)) {
+          runEff(r(value));
+        }
+        if (t = takeHead(ts)) {
+          runEff(t(value));
+        }
+      }
+      break;
+    }
+
+    // Process the next put. We do not immediately invoke the callback
+    // because we want to preserve ordering. If there are takes/reads
+    // we want to run those first.
+    if (value === EMPTY && (p = takeHead(ps))) {
+      avar.value = value = p.value;
+    }
+
+    if (value !== EMPTY) {
+      // We go ahead and queue up the next take for the same reasons as
+      // above. Invoking the read callbacks can affect the mutable queue.
+      t = takeHead(ts);
+      // We only want to process the reads queued up before running these
+      // callbacks so we guard on rsize.
+      while (rsize-- && (r = takeHead(rs))) {
+        runEff(r(right(value)));
+      }
+      if (t !== null) {
+        avar.value = EMPTY;
+        runEff(t(right(value)));
+      }
+    }
+
+    if (p !== null) {
+      runEff(p.cb(right(void 0)));
+    }
+
+    // Callbacks could have queued up more items so we need to guard on the
+    // actual mutable properties.
+    if (avar.value === EMPTY && ps.size === 0 || avar.value !== EMPTY && ts.size === 0) {
```

Comment:
> but I can't convince myself I couldn't get into a situation where it's possible to try to revoke a pending item when it hasn't been put in the appropriate queue yet

Thinking about this again, this is not a reasonable concern. If the queue is flushed, the clearly it will be in the appropriate place before yielding the canceler. Mutation hurts my brain.